src/sources/util/tcp.rs

```diff
@@ -221,59 +226,67 @@ async fn handle_stream<T>(
         }
     }
 
-    let mut shutdown_token = None;
-    let mut reader = FramedRead::new(socket, source.decoder());
-
-    stream::poll_fn(|cx| {
-        if shutdown_token.is_none() {
-            match shutdown_signal.poll_unpin(cx) {
-                Poll::Ready(token) => {
-                    debug!("Start graceful shutdown.");
-                    // Close our write part of TCP socket to signal the other side
-                    // that it should stop writing and close the channel.
-                    let socket = reader.get_ref().get_ref();
-                    if let Some(stream) = socket {
-                        let socket = SockRef::from(stream);
-                        if let Err(error) = socket.shutdown(std::net::Shutdown::Write) {
-                            warn!(message = "Failed in signalling to the other side to close the TCP channel.", %error);
+    let reader = FramedRead::new(socket, source.decoder());
+
+    let mut shutdown_signal = shutdown_signal.fuse();
+    let mut reader = reader.fuse();
+    let mut tripwire = tripwire.fuse();
+    loop {
+        futures::select! {
+            _ = tripwire => break,
+            _ = shutdown_signal => {
+                debug!("Start graceful shutdown.");
+                // Close our write part of TCP socket to signal the other side
+                // that it should stop writing and close the channel.
+                let socket = reader.get_ref().get_ref();
+                if let Some(stream) = socket.get_ref() {
+                    let socket = SockRef::from(stream);
+                    if let Err(error) = socket.shutdown(std::net::Shutdown::Write) {
+                        warn!(message = "Failed in signalling to the other side to close the TCP channel.", %error);
+                    }
+                } else {
+                    // Connection hasn't yet been established so we are done here.
+                    debug!("Closing connection that hasn't yet been fully established.");
+                    break;
+                }
+            },
+            res = reader.next() => {
+                match res {
+                    Some(Ok(frame)) => {
+                        let host = host.clone();
+                        let ack = source.build_ack(&frame);
+
+                        if let Some(event) = source.build_event(frame, host) {
+                            match out.send(event).await {
+                                Ok(_) => {
+                                    let stream = reader.get_mut().get_mut();
+                                    if let Err(err) = stream.write_all(&ack).await {
+                                        warn!("Failed to write ack: {}", err); // TODO internal event?
```

Comment:
An internal event seems prudent here, is there anything from the end-to-end work that we could borrow here?

Reply:
And of course, the `.` at the end of the message.

Reply:
Oops! Thanks. I missed this TODO when looking it over. I'll add an event.

Reply:
Added in https://github.com/timberio/vector/pull/7857/commits/7ebf1e28bb8e16c39cd08c969902a91a1a39d065